binding.c

```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
+  struct async_pwhash_str_request * req = (async_pwhash_str_request *)req_v;
+  req->n = crypto_pwhash_str(req->output_data, req->pwd_data, req->pwd_size, req->opslimit, req->memlimit);
+}
+
+static void async_pwhash_str_complete(napi_env env, napi_status status, void* data) {
+  async_pwhash_str_request * req = (async_pwhash_str_request *)data;
+  SN_STATUS_THROWS_VOID(status, "");
+
+  napi_value global;
+  SN_STATUS_THROWS_VOID(napi_get_global(env, &global), "");
+
+  napi_value argv[1];
+  napi_get_null(env, &argv[0]);
+
+  napi_value callback;
+  SN_STATUS_THROWS_VOID(napi_get_reference_value(env, req->cb, &callback), "");
+
+  napi_value return_val;
+  SN_STATUS_THROWS_VOID(napi_call_function(env, global, callback, 1, argv, &return_val), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->cb), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->output_ref), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->pwd_ref), "");
+  SN_STATUS_THROWS_VOID(napi_delete_async_work(env, req->task), "");
+  free(req);
+}
+
+napi_value sn_crypto_pwhash_str_async (napi_env env, napi_callback_info info) {
+  SN_ARGV(5, crypto_pwhash_str_async)
+
+  SN_ARGV_BUFFER_CAST(char *, output, 0)
+  SN_ARGV_BUFFER_CAST(char *, pwd, 1)
+  SN_ARGV_UINT64(opslimit, 2)
+  SN_ARGV_UINT64(memlimit, 3)
+  napi_value cb = argv[4];
+
+  SN_THROWS(output_size != crypto_pwhash_STRBYTES, "output must be 128 bytes")
+  SN_THROWS(opslimit < crypto_pwhash_OPSLIMIT_MIN, "opslimit must be at least 1")
+  SN_THROWS(opslimit > crypto_pwhash_OPSLIMIT_MAX, "opslimit must be at most 4294967295")
+  SN_THROWS(memlimit < crypto_pwhash_MEMLIMIT_MIN, "memlimit must be at least 8 kB")
+  SN_THROWS(memlimit > (int64_t) crypto_pwhash_MEMLIMIT_MAX, "memlimit must be at most 4398 GB")
+
+  async_pwhash_str_request * req = (async_pwhash_str_request *) malloc(sizeof(async_pwhash_str_request));
+  req->output_data = output;
+  req->pwd_data = pwd;
+  req->pwd_size = pwd_size;
+  req->opslimit = opslimit;
+  req->memlimit = memlimit;
+
+  SN_STATUS_THROWS(napi_create_reference(env, cb, 1, &req->cb), "")
+  SN_STATUS_THROWS(napi_create_reference(env, output_argv, 1, &req->output_ref), "")
+  SN_STATUS_THROWS(napi_create_reference(env, pwd_argv, 1, &req->pwd_ref), "")
+
+  napi_value async_resource_name;
+  SN_STATUS_THROWS(napi_create_string_utf8(env, "create_pwhash_str_async", NAPI_AUTO_LENGTH, &async_resource_name), "")
+
+  napi_create_async_work(env, NULL, async_resource_name,
+                                    async_pwhash_str_execute,
+                                    async_pwhash_str_complete,
+                                    (void*)req, &req-> task);
+
+  SN_STATUS_THROWS(napi_queue_async_work(env, req->task), "")
+
+  return NULL;
+}
+
+typedef struct async_pwhash_str_verify_request {
+  napi_ref hash_ref;
+  char * hash_data;
+  napi_ref pwd_ref;
+  const char * pwd_data;
+  size_t pwd_size;
+ 
```

Comment:
for all the _init functions we need to expose the sizeof(struct) they need so you know in JS how much to alloc.

can you add those also?

```
SN_EXPORT_UINT32(crypto_hash_sha256_sizeof, sizeof(the-struct))
```

Reply:
sorry, saw them below